### PR TITLE
Add support for Map to load from a generator 

### LINKED ISSUE #7022: Map factory does not support `pathlib.Path.glob`
### Describe the feature

`pathlib.Path.glob` produces a generator for the list of files that match the glob pattern in a particularly directory. However, this not a supported pattern for the map factory. This was surprising to me since we support `pathlib.Path` objects as well as more traditional glob patterns as well as lists.

As an example, I would expect the following to work,

```python
import pathlib
import sunpy.map
sunpy.map.Map(pathlib.Path.home().glob('*.fits'))
```
but instead it throws the following exception
```python-traceback
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [16], line 1
----> 1 sunpy.map.Map(pathlib.Path.home().glob('*.fits'))

File ~/Documents/codes/sunpy/sunpy/map/map_factory.py:254, in MapFactory.__call__(self, composite, sequence, silence_errors, *args, **kwargs)
    228 def __call__(self, *args, composite=False, sequence=False, silence_errors=False, **kwargs):
    229     """ Method for running the factory. Takes arbitrary arguments and
    230     keyword arguments and passes them to a sequence of pre-registered types
    231     to determine which is the correct Map-type to build.
   (...)
    252     as `memmap` for FITS files.
    253     """
--> 254     data_header_pairs = self._parse_args(*args, silence_errors=silence_errors, **kwargs)
    255     new_maps = list()
    257     # Loop over each registered type and check to see if WidgetType
    258     # matches the arguments.  If it does, use that type.

File ~/Documents/codes/sunpy/sunpy/map/map_factory.py:180, in MapFactory._parse_args(self, silence_errors, *args, **kwargs)
    178 for arg in args:
    179     try:
--> 180         data_header_pairs += self._parse_arg(arg, **kwargs)
    181     except NoMapsInFileError as e:
    182         if not silence_errors:

File ~/Documents/codes/sunpy/sunpy/util/functools.py:18, in seconddispatch.<locals>.wrapper(*args, **kwargs)
     17 def wrapper(*args, **kwargs):
---> 18     return dispatcher.dispatch(args[1].__class__)(*args, **kwargs)

File ~/Documents/codes/sunpy/sunpy/map/map_factory.py:195, in MapFactory._parse_arg(self, arg, **kwargs)
    189 @seconddispatch
    190 def _parse_arg(self, arg, **kwargs):
    191     """
    192     Take a factory input and parse into (data, header) pairs.
    193     Must return a list, even if only one pair is returned.
    194     """
--> 195     raise ValueError(f"Invalid input: {arg}")

ValueError: Invalid input: <generator object Path.glob at 0x104dc71b0>
```

### Proposed solution

Add this as a possible input to the map factory.

### PR DESCRIPTION
**Pull request by Jett-Code**
## PR Description
Fixes #7022

### CODE REVIEW

**Comment by nabobalis** [comment]
This will need a unit test as well to check that pathlib.glob works.

**Comment by Jett-Code** [comment]
> Looks like this change has failed on the CI, this will need debugging.
> 
> This will need a unit test as well to check that pathlib.glob works.

Could you please guide me on how to write unit tests for this?

**Comment by Jett-Code** [comment]
> You will want to add extra unit tests to https://github.com/sunpy/sunpy/blob/main/sunpy/util/tests/test_util.py to ensure that we skip over strings correctly.

There are already two unit test cases for string, list, and tuples. Do you want me to add extra on top of that?
`lst = ['a', 'b', [], (['c', 'd']), tuple(), ['e']]`

**Comment by nabobalis** [comment]
> There are already two unit test cases for string, list, and tuples. Do you want me to add extra on top of that? `lst = ['a', 'b', [], (['c', 'd']), tuple(), ['e']]`

If the tests for those are already there, then its ok. But what if you just pass in a string?

The other unit tests are failing with these changes, we need to debug this.

File: sunpy/map/tests/test_map_factory.py
Comment (nabobalis) [inline sunpy/map/tests/test_map_factory.py:90]: Why does this need to be wrapped a in list first?

Comment (Jett-Code) [reply sunpy/map/tests/test_map_factory.py:90]: Yeah! got it will correct it ASAP

**Comment by Jett-Code** [comment]
>

File: sunpy/map/tests/test_map_factory.py
Comment (Jett-Code) [reply sunpy/map/tests/test_map_factory.py:90]: I have got a small question. Which class does this object belong to? list?

Comment (nabobalis) [reply sunpy/map/tests/test_map_factory.py:90]: Which object?

Comment (PaulJWright) [reply sunpy/map/tests/test_map_factory.py:90]: Have you had the time to correct this?

**Review by PaulJWright — CHANGES_REQUESTED** [review]

File: sunpy/util/util.py
Comment (PaulJWright) [inline sunpy/util/util.py:11]: This can be removed as it is no longer needed (causing `pre-commit.ci` to fail)

**Comment by nabobalis** [comment]
Hello @Jett-Code, sorry for not getting back to you after a long time. WOuld you have some time soon to add a few unit tests for this?

**Comment by Jett-Code** [comment]
> Hello @Jett-Code, sorry for not getting back to you after a long time. WOuld you have some time soon to add a few unit tests for this?

Sure, on it!

**Comment by nabobalis** [comment]
Actually I think that one glob test is enough, I don't know why I asked for more.

Could you rebase this and we will get this reviewed!

**Review by nabobalis — APPROVED** [review]

File: sunpy/util/util.py
Comment (nabobalis) [inline sunpy/util/util.py:101]: This has to be wrong.

Comment (Jett-Code) [reply sunpy/util/util.py:101]: Got it!

File: sunpy/util/util.py
Comment (nabobalis) [inline sunpy/util/util.py:104]: A lot of extra are Iterable, there has to be a better class to avoid forming this list.

Comment (Jett-Code) [reply sunpy/util/util.py:104]: Instead of checking if the input is an instance of a specific set of types, I can check if it is an instance of the `collections.abc.Collection` class, right?

Comment (nabobalis) [reply sunpy/util/util.py:104]: Maybe that would work?

File: sunpy/util/util.py
Comment (nabobalis) [inline sunpy/util/util.py:100]: Version 3

File: sunpy/util/util.py
Comment (Jett-Code) [reply sunpy/util/util.py]: Are there any ways to find out if it's working or not? I'm confused

Comment (nabobalis) [reply sunpy/util/util.py]: The unit tests.

File: sunpy/util/util.py
Comment (Jett-Code) [reply sunpy/util/util.py:100]: It's good if it's working, right? or I'm missing something

Comment (nabobalis) [reply sunpy/util/util.py:100]: Understanding why it works is what I need to learn.

File: sunpy/util/util.py
Comment (nabobalis) [inline sunpy/util/util.py:102]:
```suggestion
        if isinstance(item, include_union):
```
Would this be enough?

Comment (Jett-Code) [reply sunpy/util/util.py:102]: I don't think so, what if someone passes a byte string?

Comment (nabobalis) [reply sunpy/util/util.py:102]: Would that not be caught by this check? 

Might be worth adding some unit tests to check `expand_list_generator` behavior?

Comment (Jett-Code) [reply sunpy/util/util.py:102]: We are checking for the list which in turn checks for the generator as input which solves the problem but here I'm adding an extra check for byte string

Comment (nabobalis) [reply sunpy/util/util.py:102]: But why would a byte string hit either of those categories?

File: sunpy/util/util.py
Comment (Jett-Code) [reply sunpy/util/util.py]: > While I doubt someone will pass a byte string in the context of this MR, as this is a utility function, I believe you want to avoid an issue with byte strings. In the current implementation a byte string wouldn't be caught by `isinstance(item, str)` as a regular string or raw string literal would be.

As suggested by @PaulJWright

File: sunpy/util/util.py
Comment (Jett-Code) [reply sunpy/util/util.py:102]: > While I doubt someone will pass a byte string in the context of this MR, as this is a utility function, I believe you want to avoid an issue with byte strings. In the current implementation a byte string wouldn't be caught by `isinstance(item, str)` as a regular string or raw string literal would be.

As suggested by @PaulJWright

Comment (nabobalis) [reply sunpy/util/util.py:102]: Sure, but I mean why would a byte string now trigger this line of code?

Comment (Jett-Code) [reply sunpy/util/util.py:102]: It triggers exclude_union which in turn triggers a string or byte

Comment (Jett-Code) [reply sunpy/util/util.py:102]: I guess we don't want any unit tests for byte strings

Comment (nabobalis) [reply sunpy/util/util.py:102]: But in this suggestion, we only check for include_union which is list, tuple and Iterator. 

I would not have assumed a byte string would trigger the last one but I don't know what Iterator fully covers.

Comment (nabobalis) [reply sunpy/util/util.py:102]: Why would we not want a unit test for it?

Comment (nabobalis) [reply sunpy/util/util.py:102]: We should expand the unit test to see what happens if we use a numpy array, a fits header, a astropy WCS, a byte string and maybe some others?

Comment (Jett-Code) [reply sunpy/util/util.py:102]: Yeah, this suggestion doesn't covers exclude_union.

Comment (Jett-Code) [reply sunpy/util/util.py:102]: > Why would we not want a unit test for byte strings?

If we have to remove byte dependency then there is no need of byte test cases

Comment (nabobalis) [reply sunpy/util/util.py:102]: > Yeah, this suggestion doesn't covers exclude_union.

Yeah, since we I don't we would need it.

Comment (nabobalis) [reply sunpy/util/util.py:102]: > > Why would we not want a unit test for byte strings?
> 
> If we have to remove byte dependency then there is no need of byte test cases

I am not sure I follow?

Comment (Jett-Code) [reply sunpy/util/util.py:102]: Ah! Never mind. I'll add unit tests for the data types and objects mentioned above

Comment (nabobalis) [reply sunpy/util/util.py]: I updated the test and the code, see what you think.

Comment (Jett-Code) [reply sunpy/util/util.py]: Seems good. What about Astropy WCS object and numpy, should I add them?

Comment (nabobalis) [reply sunpy/util/util.py]: Those should be covered by the .data and .wcs from the sunpy map.

File: sunpy/util/tests/test_util.py
Comment (nabobalis) [inline sunpy/util/tests/test_util.py:57]: Oh I should check a file handle?

Comment (Jett-Code) [reply sunpy/util/tests/test_util.py:57]: Yeah, you can exclude them from being yielded or flattened by adding checks

File: sunpy/util/util.py
Comment (Jett-Code) [inline sunpy/util/util.py:100]: This syntax is not supported for earlier versions of python, should I refactor it?

Comment (Jett-Code) [reply sunpy/util/util.py:100]: We don't want to flattened strings or byte string, that's why we have used `not isinstance(item, (bytes, str))` so that it goes through `yield item` and not through `yield from expand_list_generator(item)`

Comment (nabobalis) [reply sunpy/util/util.py:100]: > This syntax is not supported for earlier versions of python, should I refactor it?

sunpy main supports python 3.10 or above, I don't think this should be a problem.

> We don't want to flattened strings or byte string, that's why we have used `not isinstance(item, (bytes, str))` so that it goes through `yield item` and not through `yield from expand_list_generator(item)`

Oh I didn't add a test with a longer string, I should add that to check.

File: sunpy/util/tests/test_util.py
Comment (nabobalis) [inline sunpy/util/tests/test_util.py:72]: This is what I came up with, I hate it.

Comment (Jett-Code) [reply sunpy/util/tests/test_util.py:72]: Isn't this a more efficient way of doing the same thing 
```
with open(get_test_filepath('aia_171_level1.fits'), 'rb') as f:
        file_contents = f.readlines()
        assert list(util.expand_list_generator([f])) == file_contents
```

Comment (nabobalis) [reply sunpy/util/tests/test_util.py:72]: I thought that when you call readlines, it consumes f, so it's now at the end of the file. 

I guess I can just rewind it?

**Comment by Jett-Code** [comment]
Hey, can you help me with failing test cases?

**Comment by nabobalis** [comment]
> Hey, can you help me with failing test cases?

You mean the two failing checks? What would you like help with? 

The gallery one looks like a real failure due to this code. 
Unsure about the online one.

We just need to run those failing examples and debug why they are failing now and see if they are due to the code changes in this pull request.

**Comment by Jett-Code** [comment]
The examples mentioned under unexpected failing examples are working perfectly fine when executed locally, is there any other way of finding the error?

**Comment by nabobalis** [comment]
> Hey, can you help me with failing test cases?

Looks like it was due to me removing UserList, adding it back fixed it (at least the examples)

File: sunpy/util/tests/test_util.py
Comment (nabobalis) [inline sunpy/util/tests/test_util.py:66]: I assume this was not meant to happen?

**Comment by nabobalis** [comment]
The last set of pushes broke the pull request, can you revert it or fix it please?

File: sunpy/util/tests/test_util.py
Comment (Jett-Code) [reply sunpy/util/tests/test_util.py:66]: Yeah, working on merge conflicts

Comment (nabobalis) [reply sunpy/util/tests/test_util.py:66]: What caused the merge conflicts?

Comment (Jett-Code) [reply sunpy/util/tests/test_util.py:66]: last commit

**Comment by Jett-Code** [comment]
> The last set of pushes broke the pull request, can you revert it or fix it please?

Is it possible to delete the last commit?

**Comment by nabobalis** [comment]
> > The last set of pushes broke the pull request, can you revert it or fix it please?
> 
> Is it possible to delete the last commit?

Do we not need the changes in that commit?

**Comment by nabobalis** [comment]
The entire commit history is a bit strange right now. Seems to be repeated?

**Comment by nabobalis** [comment]
I pushed from an old branch I had locally. Not sure if it was up to date tho.

**Comment by Jett-Code** [comment]
> I pushed from an old branch I had locally. Not sure if it was up to date tho.

Thanks, let me check

**Comment by Jett-Code** [comment]
I guess everything is working fine now

**Comment by nabobalis** [comment]
> Is there anything else to work on regarding this pr?

No, it needs reviews from the other mainainters before a merge.

File: sunpy/util/tests/test_util.py
Comment (wtbarnes) [inline sunpy/util/tests/test_util.py:58]: This test could be cleaned up quite a bit using `parametrize`.

**Review by wtbarnes — APPROVED** [review]

File: sunpy/util/tests/test_util.py
Comment (nabobalis) [reply sunpy/util/tests/test_util.py]: Have a look now.

Comment (wtbarnes) [reply sunpy/util/tests/test_util.py]: Beautiful. Thanks!

**Comment by nabobalis** [comment]
Thanks for the PR @Jett-Code, sorry it took so long on our side.

**Pull request merged**
